enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+in the case of AWS) [https://aws.amazon.com/route53/]. OpenShift will continues to
+configure and manage the cloud load balancing solution (Elastic
+Load Balaner)[https://aws.amazon.com/elasticloadbalancing/] for all its services.
+
+The customer is also resposible for providing the OpenShift installer with an unused
+IP in the machine network for use as the API-Int server address if mulriple subnets are
+not in use. When multiple subnets are configured within the machine network, one IP
+address per subnet needs to be set aside and provided for Internal API server configuration.
+Using these IP address, OpenShift will start the CoreDNS service for API-Int server
+resolution within the cluster.
+
+The use of CoreDNS for API-Int server resolution for this solution resembles what is
+already in use for on-prem platforms but differs from them in its use of the cloud
+load balancer.
+
+### Workflow Description
+
+1. For an OpenShift administrator to use a custom DNS solution rather than
+letting OpenShift automatically configure the cloud provided, DNS solution,
+they would have to do 2 things:
+  - Pre-configure entries in their custom DNS soltion with entries for the
+API and Ingress services
+  - Let the OpenShift Installer know that they would like to use a custom
+DNS solution and skip configuration of the cloud provided DNS solution.
+
+2. The Installer next needs to validate the presence of this custom DNS and
+its state by using an utility like `dig` to determine if the API service is
+reachable and its IPs are within each of the public subnets provided as part
+of installer configuration (also called as install-config). [TODO: can we use
+teh same validation method for the Ingress service configuration?]
+
+3. When the custom DNS configuration mentioned in #1 is present, skip generating
+terraform configuration for DNS configuration in Route53.
+
+4. The Ingress Controller configures and manages the DNS configuration for the
+Ingress service. The Ingress Controller needs to be made aware of the presence
+of the custom DNS via the DNS manifest so that it skips configuring the cloud
+DNS for the Ingress service.
+
+5. The Ingress Controller continues managing the Load Balancer configuration for
+the Ingress service.
+
+6. When the custom DNS solution is enabled, installer tries to resolve the
+API Server URL to obtain the IP adderesses within each subnet in the machine
+network and validate them.
+
+7. The API-Int Server would be configured to resolve to the IP addresses provided
+by the user. (It could be configured to resolve to the same IP addresses
+as the API Server.) It is the responsibilty of the Installer to provide these
+IP addresses to the machine-config-operator (MCO) via the Infrastructure CR.
+
+8. The Installer adds the above API-Int Server IPs to the AWSPlatformStatus
+struct within the Infrastructure Status. A new field needs to be added to
+AWSPlatformStatus for this.
```

Comment:
We will need to carefully consider whether having IP addresses rather than the load balancer dns name will be sufficient. This comes up for both the external and internal load balancers.

Reply:
Yes, I was struggling with that. LB DNS names seems like an internal detail of the installation. Not sure if there is precedence for getting that type of config from the user.

Reply:
Generating the manifest within the MCO for starting the CoreDNS pod needs the values of the IP addresses to be present in the Infrastructure CR.
If the comment is about getting LB DNS names from the user via install-config, then we are also choosing the path of asking the user to pre configure their LB before calling "create cluster". So, as soon as the DNS and LB configuration workflow (as detailed in the "Design Details" section) is chosen, the install-config inputs will become clear.